### PR TITLE
testing: add content for FiveAM

### PR DESCRIPTION
**Pull request by sheepduke**
I drafted a chapter for FiveAM. I will write Mockingbird part next week when I have time.

for #297

### CODE REVIEW

**Comment by vindarel** [comment]
Yipee, thanks for tackling this issue. I think you can enable GitHub pages on your fork, so we could read this chapter for real and see how it looks? (like on https://vindarel.github.io/cl-cookbook/)

File: testing.md [hunk 1]
```diff
@@ -7,22 +7,218 @@ recipes cover how to write automated tests and see their code
...
+~~~
+
+## Mocking with Mockingbird
+
+TODO
 
 ## Testing with Prove
 
+> Warning: Prove is obsolete. This section is kept but it is not encouraged to use it.
+
```
Comment (vindarel) [inline testing.md:221]: It's OK to remove it, isn't it?

**Comment by vindarel** [comment]
It would be great if you could show how to run a test from a terminal… and get a correct exit code. I did it sometimes, I have a short snippet somewhere. There is also https://github.com/rpgoldman/fiveam-asdf (didn't use).

And mentioning https://github.com/tdrhq/slite will be cool.

File: testing.md (hunk 1 above)
Comment (sheepduke) [reply testing.md:221]: > Yipee, thanks for tackling this issue. I think you can enable GitHub pages on your fork, so we could read this chapter for real and see how it looks? (like on https://vindarel.github.io/cl-cookbook/)

Oh! Okay..I will try that :-)

**Comment by sheepduke** [comment]
> Yipee, thanks for tackling this issue. I think you can enable GitHub pages on your fork, so we could read this chapter for real and see how it looks? (like on https://vindarel.github.io/cl-cookbook/)

Oh! Okay. I will try that. :-)

File: testing.md (hunk 1 above)
Comment (sheepduke) [reply testing.md:221]: I am not sure about this. I think Prove and Rove shares many similar (or even same?) APIs. Maybe it should be somehow transformed to Rove intro?

I have not used Rove for years and do not know its status now.

What do you think?

File: testing.md [hunk 2]
```diff
@@ -7,22 +7,218 @@ recipes cover how to write automated tests and see their code
...
+(test read-file-as-string-normal-file
+  (let ((result (read-file-as-string "/tmp/hello.txt")))
+    ;; Convention: put expected value as the first arg of =, or equal, string= etc.
+    ;; FiveAM generates a more readable report following this convention.
+    (is (string= "hello" result))))
+~~~
+
+In the above code, 3 test was defined with 5 checks in total. Some checks are actually redundant for the sake of demonstration. You may put all the checks in one big test, or in multiple scenarios. It is up to you.
+
+The macro `test` is a convenience for `def-test` to define simple tests. You may read its docstring for a more complete introduction, but please note that it is not recommended to use its fixture feature according to [this issue](https://github.com/lispci/fiveam/issues/31).
```
[the remark this answers is not in the document]
Comment (sheepduke) [reply testing.md:111]: Wait..If you look at the issue, sionescu who is a contributor said:

>I'm afraid fixtures are not the best designed feature if FiveAM. In practice, I think it's simpler to just use a macro.

File: testing.md [hunk 3]
```diff
@@ -7,22 +7,218 @@ recipes cover how to write automated tests and see their code
...
+FiveAM does not have many documentation resources, yet [this blogpost](http://turtleware.eu/posts/Tutorial-Working-with-FiveAM.html) provides a detailed introduction to it. Here we will introduce it from another perspective.
 
+## Testing with FiveAM
 
-<a name="install"></a>
+FiveAM has 3 levels of abstraction: check, test and suite. As you may have guessed:
+1. A check is a single assertion that asserts a boolean value. 
+2. A test is the smallest runnable unit. A test case that may contain multiple checks. Any check failure leads to failure of the whole test.
+3. A suite is a collection of tests. When a suite is run, all tests inside would be performed. Suite allows paternity, which means that running a suite will run all the tests defined in it and in its children suites.
+
```
[the remark this answers is not in the document]
Comment (sheepduke) [reply testing.md:20]: Hmmm...Currently the code snippet is put afterwards. What kind of code example you think will be suitable here?

File: testing.md [hunk 4]
```diff
@@ -7,22 +7,218 @@ recipes cover how to write automated tests and see their code
...
+Here a new suite named `read-file-as-string` has been defined. It is declared to be a child suite of `my-system` as specified by the `:in` keyword. The macro `in-suite` sets it as the default suite for test defined later.
+
+### Defining tests
+
+The macro `test` provides a simple way to define a test with given name:
+
+~~~lisp
+(test read-file-as-string-non-existing-file
+  ;; IS accepts a boolean expression with optional reason.
```
[the remark this answers is not in the document]
Comment (sheepduke) [reply testing.md:82]: What about briefly mentioning check functions including `is`, `signals` etc? Then the comments can be shortened.

File: testing.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sheepduke) [reply testing.md:20]: Does it work now?

File: testing.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sheepduke) [reply testing.md:99]: Fixed.

**Comment by sheepduke** [comment]
> It would be great if you could show how to run a test from a terminal… and get a correct exit code. I did it sometimes, I have a short snippet somewhere. There is also https://github.com/rpgoldman/fiveam-asdf (didn't use).
> 
> And mentioning https://github.com/tdrhq/slite will be cool.

I have added introduction to Slite.

What do you mean by `run a test from a terminal`? Could you please elaborate it?

**Comment by vindarel** [comment]
Quick note: to get the correct test result when running on the terminal or on CI, I had to capture the test result myself. So inside a `run-tests.lisp` file:

```lisp
(let ((result (run!)))
  (cond
    ((null result)
     (log:info "Tests failed!")
     (uiop:quit 1))
    (t
     (log:info "All pass")
     (uiop:quit))))
```

Now Gitlab CI shows a green badge, `echo $?` shows 0 on successful tests and 1 on errors.

There might be a cleaner way with ASDF integration (and fiveam-asdf?). I wonder how others do.

**Comment by sheepduke** [comment]
> There might be a cleaner way with ASDF integration (and fiveam-asdf?). I wonder how others do.
>
Is it really a good idea to put `(uiop:quit)` in the ASDF system? It is more suitable in the CI I suppose.

IMHO it would be better to put this part of content in CI/CD section and make it more generic...

File: testing.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (vindarel) [reply testing.md:20]: yes thanks.

File: testing.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (vindarel) [reply testing.md:128]: What about not linking to the issue (that doesn't make a good impression) and just saying this: that fixtures are not "the best designed" and […] and we'll introduce another library, mockingbird.

File: testing.md (hunk 1 above)
Comment (vindarel) [reply testing.md:246]: No strong opinion. But Prove clearly is deprecated, its repository is archived. If you don't feel like updating the section for Rove in the same PR, we can delete the Prove section, and dig it up later© for Rove if needed.

**Comment by vindarel** [comment]
> Is it really a good idea to put `(uiop:quit)` in the ASDF system?

No, I don't mess with ASDF here, my snippet is to be used in another run-tests.lisp file. It simply checks the error code of `run!` and quits accordingly.

I think the current content is good!

File: testing.md [hunk 5]
```diff
@@ -7,22 +7,243 @@ recipes cover how to write automated tests and see their code
...
 
-We will use an established and well-designed regression testing
-framework called [Prove](https://github.com/fukamachi/prove). It is
-not the only possibility though,
-[FiveAM](http://quickdocs.org/fiveam/api) is a popular one (see
-[this blogpost](http://turtleware.eu/posts/Tutorial-Working-with-FiveAM.html) for an
-introduction) and [there are others](https://github.com/CodyReichert/awesome-cl#unit-testing) (and more again). We prefer
-`Prove` for its documentation and its **extensible reporters** (it has different
-report styles and we can extend them).
+[Prove](https://github.com/fukamachi/prove) used to be a widely liked testing framework but its repository was later archived. Its successor [Rove](https://github.com/fukamachi/rove) is not stable enough, so we will be using a mature testing framework called [FiveAM](https://github.com/lispci/fiveam). There are also some [other testing frameworks](https://github.com/CodyReichert/awesome-cl#unit-testing) to explore.
```
Comment (vindarel) [inline testing.md:10]: You don't link to 5am docs! (they are not reachable from its Github)

**Comment by vindarel** [comment]
notes:

It looks like def-suite and test accept a `description`

```lisp
(def-suite random-utils-tests
    :description "Test the random utilities."
    :in all-tests)
```

We can also nest suites with `:in`.

When we give a description, we can use format placeholders:

```lisp
(is (string= "Tom" name)
    "MAKE-CHARACTER should capitalize the name \"tom\", but we got: ~s" name))
```

thanks http://turtleware.eu/posts/Tutorial-Working-with-FiveAM.html

File: testing.md (hunk 1 above)
Comment (AlbertoEAF) [reply testing.md:246]: Hi all :)

I was looking at the page, and for complete beginners, it is more informative and gives a better outlook if we state that it's superseded by Rove, by the same author.

Something like:

> Warning: Prove is obsolete and superseded by [Rove](https://github.com/fukamachi/rove), by the same author. Both have similar APIs. This section is kept as an overview of Prove until a rewrite is done for Rove.

File: testing.md [hunk 6]
```diff
@@ -7,22 +7,243 @@ recipes cover how to write automated tests and see their code
...
+; Test READ-FILE-AS-STRING-NON-EXISTING-FILE passed => NIL
+~~~
+
+## Mocking with Mockingbird
+
+TODO
 
 ## Testing with Prove
 
+> Warning: Prove is obsolete. This section is kept but it is not encouraged to use it.
```
Comment (AlbertoEAF) [inline testing.md:245]:
```suggestion
Warning: Prove is obsolete and superseded by Rove, by the same author. Both have similar APIs. This section is kept as an overview of Prove until a rewrite is done for Rove.
```

File: testing.md (hunk 5 above)
Comment (sheepduke) [reply testing.md:10]: OH! My fault. It only has a API doc that I simply missed it. :-(

File: testing.md (hunk 1 above)
Comment (sheepduke) [reply testing.md:246]: This is actually the initial intention of mine, i.e. to keep it as a reference for Rove.

But now I am concerned that their APIs are somehow different. If so, this content might be a bit misleading...

@vindarel You make the call.

**Comment by sheepduke** [comment]
> Yipee, thanks for tackling this issue. I think you can enable GitHub pages on your fork, so we could read this chapter for real and see how it looks? (like on https://vindarel.github.io/cl-cookbook/)

@vindarel I just randomly found a solution. In `Files changed` tab, there is a small page icon beside `<>` named `Display the rich diff`. GitHub will render the content and display it for you.

The downside is that you cannot comment on rich diff mode. Yet it could be solved by opening another tab for the source...

File: testing.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sheepduke) [reply testing.md]: Fixed.

File: testing.md (hunk 5 above)
Comment (sheepduke) [reply testing.md:10]: Fixed.

File: testing.md (hunk 1 above)
Comment (AlbertoEAF) [reply testing.md:248]: I think it's well written and works for Prove, which isn't that different from Rove.

If you leave it in I might update it for Rove in the next weeks ;)

Comment (vindarel) [reply testing.md:248]: If another voice says it's better for beginners to keep the Prove, so let's keep it, thanks for the feedback.

But then to properly close the issue we will need to 1) update it to Rove obviously and 2) compare the two frameworks and say what advantages there are to use Rove. The one I saw, using it from the command line, doesn't work as advertised.

**Comment by vindarel** [comment]
Hey @sheepduke, I believe we can merge this PR, if you removed the "mocking with mockingbird: todo" section, or write it. (do you plan to do it soon-ish? Your call)

**Comment by sheepduke** [comment]
> Hey @sheepduke, I believe we can merge this PR, if you removed the "mocking with mockingbird: todo" section, or write it. (do you plan to do it soon-ish? Your call)
> 
I have removed this line and will merge this PR. 

By the way, I recently saw extensive usage of fixture feature in project `cl-gserver`.
https://github.com/mdbergmann/cl-gserver/blob/master/tests/actor-test.lisp

Maybe it is not that bad. I will test it more and modify it if needed when adding Mockingbird thing.

**Pull request merged**